source/linked_list.c

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/linked_list.h>
+#include <string.h>
+
+void aws_linked_list_init(struct aws_linked_list_node *head) {
+    head->next = head;
+    head->prev = head;
+}
+
+void aws_linked_list_clear(struct aws_linked_list_node *head) {
+    head->next = head;
+    head->prev = head;
+}
+
+static void aws_linked_list_add(struct aws_linked_list_node *new,
+        struct aws_linked_list_node *prev,
+        struct aws_linked_list_node *next) {
+
+    next->prev = new;
+    new->next = next;
+    new->prev = prev;
+    prev->next = new;
+}
+
+void aws_linked_list_insert_before(struct aws_linked_list_node *head, struct aws_linked_list_node *new) {
```

Comment:
You could also implement this as:

    aws_linked_list_insert_after(head->prev, new);

This eliminates the need for a helper function, which can be useful if you're going to static inline these.